File: golang/ast/path.go
package ast

import (
	"fmt"
	"strings"
)

func (p *Path) AppendIndex(i int) *Path {
	return &Path{
		Pos: append(append([]*Path_Pos{}, p.Pos...), &Path_Pos{Index: int64(i)}),
	}
}

func (p *Path) AppendKey(k string) *Path {
	return &Path{
		Pos: append(append([]*Path_Pos{}, p.Pos...), &Path_Pos{Key: k}),
	}
}

func (p *Path) Format() string {
	var s []string
	for _, pos := range p.Pos {
		if pos.Key == "" {
			s = append(s, fmt.Sprint(pos.Index))
		} else {
			s = append(s, fmt.Sprintf("%v", pos.Key))
		}
	}
	return "/" + strings.Join(s, "/")
}

File: dart/lib/eval/path.dart
import 'package:eyamluate/eval/eval.dart';
import 'package:fixnum/fixnum.dart';

Path pathAppendKey(Path path, String key) {
  return Path(pos: List.from(path.pos)..add(Path_Pos(key: key)));
}

Path pathAppendIndex(Path path, int index) {
  return Path(pos: List.from(path.pos)..add(Path_Pos(index: Int64(index))));
}
